lib/Dist/Zilla/Dist/Builder.pm

```diff
@@ -483,17 +496,21 @@ sub build_archive {
 
   my $built_in = $self->ensure_built;
 
-  my $basename = $self->dist_basename;
-  my $basedir = path($basename);
+  my $basedir = path($self->dist_basename);
 
-  $_->before_archive for @{ $self->plugins_with(-BeforeArchive) };
+  $_->before_archive for $self->plugins_with(-BeforeArchive)->@*;
+
+  for my $builder ($self->plugins_with(-ArchiveBuilder)->@*) {
+    my $file = $builder->build_archive($self->archive_basename, $built_in, $basedir);
```

Comment:
- alternatively we could pass in arguments as a  hash reference
 - could also omit `$self->archive_basename` as an argument; if this is a public method I think it can be called from it from the ArchiveBuilder

Reply:
In fact I think all of these arguments could be fetched from the `$zilla` object?